Metrics API refactor

YapingLi04 opened the pull request.

YapingLi04 committed:
[metrics] Introduce metrics API framework

This commit introduces the shared code for the metrics API framework:

- Metrics API definitions
- Code to set up the varlink server
- The describe method which shows all the metrics families
- The list method which lists all the metrics
- Type definitions related to MetricFamily
- Common code to build json objects

daandemeyer commented on src/shared/metrics.c:
```diff
@@ -0,0 +1,268 @@
+                if (r < 0)
+                        return r;
+
+                // sd_json_variant_unrefp(TAKE_PTR(&context->previous)); // Yaping: check &
```
`sd_json_variant_unref(context->previous)`

ikruglov replied:
`context->previous = sd_json_variant_unref(context->previous)` I'd say to ensure no reuse. At the same time the variable is overwriteen right below. Up to you.

daandemeyer commented on src/shared/metrics.c:
```diff
@@ -0,0 +1,268 @@
+        if (r < 0)
+                return r;
+
+        if (fields) { /* NULL => no fields object, empty strv => field: {} */
```
"fields: {}"

daandemeyer commented on src/shared/metrics.c:
```diff
@@ -0,0 +1,268 @@
+
+int metric_set_value_unsigned(sd_json_variant **v, uint64_t value) {
+        return sd_json_variant_set_field_unsigned(ASSERT_PTR(v), "value", value);
+}
```
Both are unused and can be removed

YapingLi04 replied:
Dropped and metric_build_body_json_one

daandemeyer commented on src/shared/metrics.c:
```diff
@@ -0,0 +1,268 @@
+                MetricFamilyContext ctx = {
+                        .metric_family = mf,
+                        .link = link,
+                        .previous = previous};
```
You can declare this outside of the for loop and only update the `metric_family` field inside the for loop. Then you can get rid of the `previous` variable declaration outside of the for loop.

YapingLi04 replied:
updated.

daandemeyer commented on src/shared/metrics.c:
```diff
@@ -0,0 +1,268 @@
+        }
+
+        /* produce last empty metric to notify client about more = false */
+        return sd_varlink_reply(link, NULL);
```
This needs to check if `ctx.previous` is set and if so send it without the `more` flag. The entire reason we introduced `MetricFamilyContext` is so that we wouldn't have to send an empty object at the end anymore.

YapingLi04 replied:
updated.

daandemeyer commented on src/core/varlink-metrics.c:
```diff
@@ -0,0 +1,167 @@
+                                value,
+                                STRV_MAKE("type", unit_type_to_string(type)));
+                if (r < 0)
+                        return r;
```
This needs to be switched to `metric_build_send()` as well

YapingLi04 replied:
switched and removed `metric_build_full_json_one` in all files.

daandemeyer commented on src/core/varlink-metrics.c:
```diff
@@ -0,0 +1,167 @@
+#include "unit-def.h"
+#include "varlink-metrics.h"
+
+static int unit_active_state_build_json(MetricFamilyContext *metric_family_context, void *userdata) {
```
The type name is already long enough, you can just use `context` for the function argument name instead of `metric_family_context`

YapingLi04 replied:
Updated all `metric_family_context` to `context` in all functions in this file.

daandemeyer commented on src/core/varlink-metrics.c:
```diff
@@ -0,0 +1,167 @@
+                _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+                r = sd_json_variant_new_string(&v, unit_active_state_to_string(unit_active_state(unit)));
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to create sd_json_variant_new_string for object'%s' for active state: %m", unit->id);
```
Missing whitespace after `object`

YapingLi04 replied:
object is dropped after making the comment less verbose based on your other comment.

daandemeyer commented on src/core/varlink-metrics.c:
```diff
@@ -0,0 +1,167 @@
+                _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+                r = sd_json_variant_new_string(&v, unit_load_state_to_string(unit->load_state));
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to create sd_json_variant_new_string for object '%s' for load state: %m", unit->id);
```
"Failed to allocate JSON string", no need to be very verbose with the log message here, the only reason this will ever fail is if we're out of memory.

YapingLi04 replied:
updated all similar error messages.

daandemeyer commented on src/core/varlink-metrics.c:
```diff
@@ -0,0 +1,167 @@
+                                metric_family_context->metric_family,
+                                /* object= */ NULL,
+                                value,
+                                STRV_MAKE("state", unit_active_state_to_string(state)));
```
This needs to be switched to metric_build_send() as well

YapingLi04 replied:
done

daandemeyer commented on src/shared/metrics.c:
```diff
@@ -0,0 +1,268 @@
+        }
+
+        return sd_varlink_notify(link, v);
+}
```
This can be dropped once the two functions in the next commit are switched to metric_build_send()

YapingLi04 replied:
dropped

daandemeyer commented on src/shared/metrics.c:
```diff
@@ -0,0 +1,268 @@
+                return log_debug_errno(r, "Failed to call fill_metric_cb for '%s': %m", mf->name);
+
+        return sd_varlink_notify(link, v);
+}
```
This is unused now

YapingLi04 replied:
dropped.

daandemeyer commented on src/shared/metrics.h:
```diff
@@ -0,0 +1,54 @@
+        const char *description;
+        MetricFamilyType type;
+        metric_family_iterate_cb_t iterate_cb;
+        metric_family_fill_metric_cb_t fill_metric_cb;
```
This is unused now

YapingLi04 replied:
dropped

daandemeyer commented on src/shared/metrics.h:
```diff
@@ -0,0 +1,54 @@
+} MetricFamilyContext;
+
+typedef int (*metric_family_iterate_cb_t) (MetricFamilyContext *mfc, void *userdata);
+typedef int (*metric_family_fill_metric_cb_t) (sd_json_variant **v, void *userdata);
```
This is also unused

YapingLi04 replied:
dropped

daandemeyer commented on src/shared/metrics.h:
```diff
@@ -0,0 +1,54 @@
+
+typedef struct MetricFamily MetricFamily;
+
+typedef struct MetricFamilyContext {
+        const MetricFamily* metric_family;
+        sd_varlink *link;
+        sd_json_variant *previous;
+} MetricFamilyContext;
```
Please declare this after `MetricFamilyContext` so you don't need the separate typedef

YapingLi04 replied:
@DaanDeMeyer, Did you mean to declare `MetricFamilyContext` after `MetricFamily`?

`MetricFamily` depends on `metric_family_iterate_cb_t` which depends on `MetricFamilyContext`.

daandemeyer replied:
Ah right I missed that forget about this comment then.

daandemeyer commented on src/shared/metrics.h:
```diff
@@ -0,0 +1,54 @@
+int metric_build_body_json_one(sd_varlink *link, const MetricFamily *mf, const char *object, void *userdata);
+int metric_set_value_string(sd_json_variant **v, const char *value);
+int metric_set_value_unsigned(sd_json_variant **v, uint64_t value);
+int metric_set_fields(sd_json_variant **v, char **fields);
```
This does not need to be a public function anymore

YapingLi04 replied:
made it static

daandemeyer commented on src/report/report.c:
```diff
@@ -0,0 +1,212 @@
+
+        /* This is mostly intended to be used for scripts which want
+                * to detect whether we are being run in a virtualized
+                * environment or not */
```
Copy pasted comment

YapingLi04 replied:
removed comment

daandemeyer commented on src/report/report.c:
```diff
@@ -0,0 +1,212 @@
+                return log_error_errno(r, "Failed to enable exit on SIGINT/SIGTERM: %m");
+
+        sd_varlink *varlinks[MAX_METRICS_SOCKETS] = {NULL};
+        Context context = {0};
```
No need for the zero

daandemeyer reviewed (changes requested):
Can you split the commit introducing the metrics and the report CLI into two separate commits?

ikruglov commented on src/core/varlink-metrics.c:
```diff
@@ -0,0 +1,167 @@
+                _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+                r = sd_json_variant_new_string(&v, unit_active_state_to_string(unit_active_state(unit)));
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to create sd_json_variant_new_string for object'%s' for active state: %m", unit->id);
```
please create type-specific functions wraping `metric_build_send()`

YapingLi04 replied:
@ikruglov Could you please elaborate what you mean by creating type-specific functions wraping metric_build_send()?

ikruglov replied:
create functions like:
```
int metric_build_send_unsigned(MetricFamilyContext* context, const char *object, uint64_t value, char **fields);
int metric_build_send_string(MetricFamilyContext* context, const char *object, const char *value, char **fields);
etc
```

so you can simplify code by avoiding jsonizing the value manually

YapingLi04 replied:
done.

ikruglov commented on src/core/varlink-metrics.c:
```diff
@@ -0,0 +1,167 @@
+static int unit_active_state_build_json(MetricFamilyContext *metric_family_context, void *userdata) {
+        Manager *manager = ASSERT_PTR(userdata);
+        Unit *unit;
+        int r;
```
missing assert

ikruglov replied:
in other places as well

YapingLi04 replied:
added `assert(context)` in relevant functions in this file

ikruglov commented on src/core/varlink-metrics.c:
```diff
@@ -0,0 +1,167 @@
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to create sd_json_variant_new_string for object'%s' for active state: %m", unit->id);
+
+                r = metric_build_send(metric_family_context, unit->id, v, /* fields */ NULL);
```
/* fields = */

YapingLi04 replied:
updated other places as well.

ikruglov commented on src/core/varlink-metrics.c:
```diff
@@ -0,0 +1,167 @@
+        return 0;
+}
+
+static int units_by_type_total_build_json(MetricFamilyContext *metric_family_context, void *userdata) {
```
make sure that existing metric functions are compatible with next setup using family context.

YapingLi04 replied:
@ikruglov did you mean to replace `metric_build_full_json_one` with `metric_send_build` in this file as Daan commented below?

ikruglov replied:
yes, basically

YapingLi04 replied:
That's done.

ikruglov commented on src/core/varlink.c:
```diff
@@ -1,10 +1,13 @@
 /* SPDX-License-Identifier: LGPL-2.1-or-later */
 
+#include <assert.h>
```
is there a need to include assert?

YapingLi04 replied:
dropped.

ikruglov commented on src/core/varlink.c:
```diff
@@ -420,6 +424,43 @@ int manager_setup_varlink_server(Manager *m) {
+                        return r;
+
+                if (!user_address)
+                        return -ENOMEM;
```
ensure this check is not redundant

YapingLi04 replied:
@ikruglov `user_address` is the `ret` parameter in `runtime_directory_generic()`. The check here ensure that `user_address` is assigned to a non NULL value.

`int runtime_directory_generic(RuntimeScope scope, const char *suffix, char **ret);`

ikruglov replied:
yes, but you also check exit code `r`. systemd's codebase is quite consistent of ensuring that `user_address` would have a meaningful value upon successful completion. 

This check seems redundant as result. Check implementation details of `runtime_directory_generic()` just to be sure.

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,212 @@
+
+        fflush(stdout);
+
+        if ((flags == 0) && (--context->n_ops <= 0))
```
I'm not sure if `flags == 0` is a correct check. You want to ensure that you decrement counter on the last message for each connection. Other flags may still be set.

ikruglov replied:
also, consider handling SD_VARLINK_REPLY_ERROR

YapingLi04 replied:
@ikruglov, I tested the value of `flags`. It's 0 on the last message. For all other messages, the it's `SD_VARLINK_REPLY_CONTINUES` which is 2.

This the enum.
```
__extension__ typedef enum _SD_ENUM_TYPE_S64(sd_varlink_reply_flags_t) {
        SD_VARLINK_REPLY_ERROR     = 1 << 0,
        SD_VARLINK_REPLY_CONTINUES = 1 << 1,
        SD_VARLINK_REPLY_LOCAL     = 1 << 2,
        _SD_ENUM_FORCE_S64(SD_VARLINK_REPLY)
} sd_varlink_reply_flags_t;

```
I added statement to handle `SD_VARLINK_REPLY_ERROR`.

ikruglov replied:
> For all other messages, the it's SD_VARLINK_REPLY_CONTINUES which is 2.

I think you should be checking for this flag specifically rather then doing a coarse `flags == 0`.

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,212 @@
+        if (error_id) {
+                bool disconnect;
+
+                disconnect = streq(error_id, SD_VARLINK_ERROR_DISCONNECTED);
```
coalesce these two lines

YapingLi04 replied:
done

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,212 @@
+        int r;
+
+        assert(path);
+        assert(event);
```
missing assert on vl and context

YapingLi04 replied:
@ikruglov vl is `NULL`. Added assert on context.

YapingLi04 replied:
added assert(vl).

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,212 @@
+
+        r = sd_varlink_observe(vl, "io.systemd.Metrics.List", /* parameter */ NULL);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to issue List() call: %m");
```
`Failed to issue io.systemd.Metrics.List call`

YapingLi04 replied:
updated

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,212 @@
+                        return -ENOMEM;
+
+                if (de->d_type != DT_SOCK && de->d_type != DT_UNKNOWN)
+                        continue;
```
skip first before doing `path_join()`

YapingLi04 replied:
done

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,212 @@
+                if (de->d_type != DT_SOCK && de->d_type != DT_UNKNOWN)
+                        continue;
+
+                varlinks[i] = sd_varlink_unref(varlinks[i]);
```
what is this call supposed do? `varlink[i]` is not initialized

YapingLi04 replied:
`_cleanup_(sd_varlink_unref) sd_json_variant *varlinks[MAX_CONCURRENT_METRICS_SOCKETS] = {}; gives the following error message:

error: passing argument 1 of ‘sd_varlink_unref’ from incompatible pointer type [-Wincompatible-pointer-types]
  108 |         _cleanup_(sd_varlink_unref) sd_json_variant *varlinks[MAX_CONCURRENT_METRICS_SOCKETS] = {};
      |         ^~~~~~~~~
      |         |
      |         sd_json_variant * (*)[20] {aka struct sd_json_variant * (*)[20]}
`

This is one of the solutions devmate suggested how to fix this issue. I changed the code to manually call `sd_varlink_unref` when the event loop exits.

ikruglov replied:
```
cleanup(sd_varlink_unref) sd_json_variant *varlinks[MAX_CONCURRENT_METRICS_SOCKETS] = {}; gives the following error message:

error: passing argument 1 of ‘sd_varlink_unref’ from incompatible pointer type [-Wincompatible-pointer-types]
108 | cleanup(sd_varlink_unref) sd_json_variant varlinks[MAX_CONCURRENT_METRICS_SOCKETS] = {};
```

this doesn't work because `sd_varlink_unref` expect a pointer to varlink object, not array of them.

YapingLi04 replied:
Yes. What you said is correct.

@ikruglov Please see the code where I manually call sd_varlink_unref when the event loop exits. Does it look correct to you?

YapingLi04 replied:
Oh, I saw your comment below about early returns and leaky variables.

YapingLi04 replied:
fixed leaky variables by introducing `sd_varlink_unref_many`

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,212 @@
+        if (r < 0)
+                return log_error_errno(r, "Failed to enable exit on SIGINT/SIGTERM: %m");
+
+        sd_varlink *varlinks[MAX_METRICS_SOCKETS] = {NULL};
```
this variable leaks

ikruglov replied:
also, no need for NUL. `= {}` does zero-initialization.

YapingLi04 replied:
See my comment below on line 118.

ikruglov replied:
the variable still leaks. Think about early return cases which you have between declaring array and cleaning up the array. 

This is why there are cleanup functions (ex: `cleanup(sd_varlink_unref)`) which deallocates a variable upon leaving current stack no matter where return is. I'm not sure if there is preexisting one for array of `sd_varlinks`. This is what you need.

YapingLi04 replied:
hmm, I was looking at early returns where `sd_event_exit()` are all called and the code will go here and `sd_varlink_unref` will be called.

```
        r = sd_event_loop(event);
        for (i = 0; i < MAX_CONCURRENT_METRICS_SOCKETS; i++)
                sd_varlink_unref(varlinks[i]);

```
Is there a way where we don't have to call `sd_varlink_unref` manually?

ikruglov replied:
there is `CLEANUP_ARRAY` but it needs `sd_varlink_unref_many` function which do not exist.

YapingLi04 replied:
done. thanks for the pointers!

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,212 @@
+                if (r < 0)
+                        return r;
+
+                if (++i >= MAX_METRICS_SOCKETS)
```
you can use `GREEDY_REALLOC()` for dynamic reallocation for the array.

ikruglov replied:
at the same time, setting max amount of concurrent requests is probably a good idea. But IMO we should handle first N instead of failing all N request at N+1.

YapingLi04 replied:
Updated to break instead of throw so we handle the first N requests.

ikruglov commented on src/shared/varlink-io.systemd.Metrics.c:
```diff
@@ -0,0 +1,52 @@
+                SD_VARLINK_DEFINE_FIELD(name, SD_VARLINK_STRING, SD_VARLINK_NULLABLE),
+                /* metric value has various types depending on MetricFamilyType and actual data double/int/uint */
+                SD_VARLINK_FIELD_COMMENT("Metric value"),
+                SD_VARLINK_DEFINE_FIELD(value, SD_VARLINK_OBJECT, SD_VARLINK_NULLABLE),
```
since we no longer send empty metrics `name` and `value` should not be nullable

YapingLi04 replied:
updated to 0

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,212 @@
+                const char *error_id,
+                sd_varlink_reply_flags_t flags,
+                void *userdata) {
+
```
no new line here

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,212 @@
+#define MAX_METRICS_SOCKETS 20
+
+typedef struct Context {
+        int n_ops;
```
ref counting variables are usually `unsigned` and called `n_ref`

YapingLi04 replied:
updated.

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,212 @@
+}
+
+static int parse_argv(int argc, char *argv[]) {
+
```
no empty line

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,212 @@
+        assert(argv);
+
+        while ((c = getopt_long(argc, argv, "hp", options, NULL)) >= 0)
+
```
here as well

YapingLi04 replied:
removed the new line

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,212 @@
+        while ((c = getopt_long(argc, argv, "hp", options, NULL)) >= 0)
+
+                switch (c) {
+
```
and here, and below

YapingLi04 replied:
removed unnecessary newline.

YapingLi04 reviewed (commented):
Addressed all the comments.

There are few unresolved comments which I had follow up questions or comments.

I made changes in the topmost commit and didn't do it in the two commits where the comments were. This way I can test the code still works. I will split the changes in the local units in the upstream PR.

daandemeyer commented on src/shared/metrics.c:
```diff
@@ -177,37 +176,11 @@ static int metric_set_fields(sd_json_variant **v, char **fields) {
-        }
-
-        return sd_varlink_notify(link, v);
-}
```
This was removed in the wrong commit, it is added in the first commit and then removed again in the second commit, it should be removed in the first commit.

YapingLi04 replied:
will fix move changes to the proper commit before updating the PR to upstream. 

I made the majority of the changes in the second commit so I can incrementally make changes and testing them. If I make lots of changes in the first commit (and test in the second commit) and one of them is incorrect, it's a little difficult to tell which one caused the issue.

YapingLi04 replied:
done

daandemeyer commented on src/shared/metrics.c:
```diff
@@ -221,7 +194,7 @@ int metric_build_send(MetricFamilyContext* context, const char *object, sd_json_
         if (r < 0)
                 return r;
 
-        if (fields) { /* NULL => no fields object, empty strv => field:{} */
```
Same here, you're fixing stuff from the first commit in the second commit, it should be fixed in the first commit.

YapingLi04 replied:
done

ikruglov commented on src/shared/metrics.c:
```diff
@@ -0,0 +1,214 @@
+        MetricFamilyContext ctx = {
+                .metric_family = NULL,
+                .link = link,
+                .previous = NULL};
```
you don't need to initialize fields to NULL when using `= {}` construction. It's so call null-initializer. It will write 0s to all of the fields which are not explicitly mentioned inside brackets.

```
MetricFamilyContext ctx = { .link = link };
```

YapingLi04 replied:
done.

ikruglov commented on src/shared/metrics.c:
```diff
@@ -0,0 +1,214 @@
+        if (ctx.previous)
+                return sd_varlink_reply(link, ctx.previous);
+        else
+                return sd_varlink_error(link, SD_VARLINK_ERROR_EXPECTED_MORE, NULL);
```
`SD_VARLINK_ERROR_EXPECTED_MORE` should be use when a function expect --more but it's not passed. This is not the case here. The appropriate error is NoSuchMetric here.

Also, please use an early return, instead of if/else:

```
if (!ctx.previous))
   return sd_varlink_error(link, NO_SUCH_METRIC, NULL);

return sd_varlink_reply(link, ctx.previous);
```

ikruglov replied:
another thing, `ctx.previous` leaks here. You have to unref it before returning from the function.

ikruglov replied:
unfortunately, this will require defining a custom deallocator for MetricFamilyContext. I don't see a way around it.

look for example in `src/machine/machine-varlink.c`: `machine_open_paramaters_done` and `MachineOpenParameters`

YapingLi04 replied:
`ctx` was allocated in `metrics_method_list()` on the stack. Do we need to worry about memory leaks?

YapingLi04 replied:
defined a custom deallocator for MetricFamilyContext. Thanks for the code pointer!

ikruglov commented on src/core/varlink-metrics.c:
```diff
@@ -0,0 +1,168 @@
+        assert(context);
+
+        HASHMAP_FOREACH(unit, manager->units) {
+                _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
```
you need to skip aliases. Look for this snippet in the codebase:

```
                 if (k != u->id)
                         continue;
```

ikruglov replied:
below as well

YapingLi04 replied:
updated all three `HASHMAP_FOREACH` to `HASHMAP_FOREACH_KEY`

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,208 @@
+
+        r = sd_event_loop(event);
+        for (i = 0; i < MAX_CONCURRENT_METRICS_SOCKETS; i++)
+                sd_varlink_unref(varlinks[i]);
```
as discussed, this doesn't slow the problem of leaking `varlinks` variable. Need `CLEANUP_ARRAY`

YapingLi04 replied:
defined `sd_varlink_unref_many` and used `CLEANUP_ARRAY`.

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,208 @@
+        r = sd_event_get_exit_code(event, &c);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get exit code: %m");
+
```
double empty line

YapingLi04 commented:
Added type specific `metric_build_send()`
fixed leaking varlinks array

will address remaining comment in the next around.

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,215 @@
+
+                r = metrics_call(p, event, varlinks[i], &context);
+                if (r < 0)
+                        goto finish;
```
IMO we should not abort if one socket is unresponsive.
I'd also think about timeouts.

ikruglov replied:
when you're going to continue, instead of aborting, you need to be careful with reference counting. Because in your current implementation a failing `metrics_call()` leads to never ending event loop.

YapingLi04 replied:
added timeout for the event loop. Used 30 seconds as timeout.

decremented the reference count after a failing `metrics_call()`.

ikruglov commented on src/shared/metrics.c:
```diff
@@ -0,0 +1,234 @@
+}
+
+static int metric_set_fields(sd_json_variant **v, char **fields) {
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *fields_v = NULL;
```
name the variable "v". It's a common name for short functions.

YapingLi04 replied:
We can't. There's a `v` parameter passed into the function `metric_set_fields`.

ikruglov commented on src/shared/metrics.c:
```diff
@@ -0,0 +1,234 @@
+int metric_build_send_string(MetricFamilyContext* context, const char *object, const char *value, char **fields) {
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+
+        int r = sd_json_variant_new_string(&v, value);
```
```
int r;
```

`r` should be on a separate line. Always. The maintainers are quite picky about it

YapingLi04 replied:
I purposely changed it to one line since `r` is used only once in the function.

Updated to two lines.

ikruglov commented on src/shared/metrics.c:
```diff
@@ -0,0 +1,234 @@
+}
+
+int metric_build_send_string(MetricFamilyContext* context, const char *object, const char *value, char **fields) {
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
```
assert(value);

ikruglov commented on src/shared/metrics.c:
```diff
@@ -0,0 +1,234 @@
+}
+
+int metric_build_send_unsigned(MetricFamilyContext* context, const char *object, uint64_t value, char **fields) {
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
```
same here

YapingLi04 replied:
done

ikruglov commented on src/shared/varlink-io.systemd.Metrics.c:
```diff
@@ -0,0 +1,52 @@
+               SD_VARLINK_SYMBOL_COMMENT("Method to get the metric families"),
+               &vl_method_Describe,
+               SD_VARLINK_SYMBOL_COMMENT("No such metric found"),
+               &vl_error_NoSuchMetric);
```
indentation seems off

YapingLi04 committed:
[metrics] Add basic system wide and per unit metrics

This commit adds some basic metrics and integration tests.

System wide metrics:
- units_by_type_total: target/device/automount etc.
- units_by_state_total: active/reloading/inactive etc.

Two per unit metrics which shows the current state of a unit:
- unit_active_state
- unit_load_state

A metric for service state:
- nrestarts

Here are some sample outputs:

units_by_type_total:

{
        "name" : "io.systemd.Manager.units_by_type_total",
        "value" : 52,
        "fields" : {
                "type" : "target"
        }
}
{
        "name" : "io.systemd.Manager.units_by_type_total",
        "value" : 82,
        "fields" : {
                "type" : "device"
        }
}
{
        "name" : "io.systemd.Manager.units_by_type_total",
        "value" : 2,
        "fields" : {
                "type" : "automount"
        }
}

units_by_state_total:

{
        "name" : "io.systemd.Manager.units_by_state_total",
        "value" : 216,
        "fields" : {
                "state" : "active"
        }
}
{
        "name" : "io.systemd.Manager.units_by_state_total",
        "value" : 0,
        "fields" : {
                "state" : "reloading"
        }
}
{
        "name" : "io.systemd.Manager.units_by_state_total",
        "value" : 120,
        "fields" : {
                "state" : "inactive"
        }
}

unit_active_state:

{
        "name" : "io.systemd.Manager.unit_active_state",
        "object" : "multi-user.target",
        "value" : "active"
}
{
        "name" : "io.systemd.Manager.unit_active_state",
        "object" : "systemd-sysusers.service",
        "value" : "inactive"
}

unit_load_state:

{
        "name" : "io.systemd.Manager.unit_load_state",
        "object" : "multi-user.target",
        "value" : "loaded"
}

nrestarts:

{
        "name" : "io.systemd.Manager.nrestarts",
        "object" : "user@0.service",
        "value" : 0
}
{
        "name" : "io.systemd.Manager.nrestarts",
        "object" : "user-runtime-dir@0.service",
        "value" : 0
}

YapingLi04 committed:
[metrics] Add a basic CLI

systemd-report will list all the metrics.

ikruglov commented on src/core/varlink.c:
```diff
@@ -420,6 +422,44 @@ int manager_setup_varlink_server(Manager *m) {
+                        return r;
+
+                if (!user_address)
+                        return -ENOMEM;
```
I still think this check is redundant. Please verify and remove

YapingLi04 replied:
it indeed is redundant. Check removed.
```
int runtime_directory_generic(RuntimeScope scope, const char *suffix, char **ret) {
...
        case RUNTIME_SCOPE_SYSTEM: {
                char *d = path_join("/run", suffix);
                if (!d)
                        return -ENOMEM;
                *ret = d;
                break;
        }
```

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,227 @@
+        }
+
+        if ((flags == 0) && (--context->n_ref <= 0))
+                (void) sd_event_exit(ASSERT_PTR(sd_varlink_get_event(link)), EXIT_SUCCESS);
```
@YapingLi04 did you look into checking a specific flags rather then doing a coarse `flags == 0`?

YapingLi04 replied:
changed to check `!FLAGS_SET(flags, SD_VARLINK_REPLY_CONTINUES)`

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,227 @@
+static void sd_varlink_unref_many(sd_varlink **array, size_t n) {
+        FOREACH_ARRAY(v, array, n)
+                sd_varlink_unref(*v);
+
```
double empty line

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,227 @@
+}
+
+static void sd_varlink_unref_many(sd_varlink **array, size_t n) {
+        FOREACH_ARRAY(v, array, n)
```
missing asserts

YapingLi04 replied:
added

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,227 @@
+}
+
+static int timeout_callback(sd_event_source *es, uint64_t usec, void *userdata) {
+        return sd_event_exit(ASSERT_PTR(sd_event_source_get_event(es)), EXIT_FAILURE);
```
worth logging that request(s) timed out

YapingLi04 replied:
done

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,227 @@
+                if (errno == ENOENT)
+                        return -ESRCH;
+
+                return -errno;
```
please log here

YapingLi04 replied:
done

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,227 @@
+                if (!p)
+                        return -ENOMEM;
+
+                context.n_ref++;
```
why not incrementing after the `metrics_call()`?

YapingLi04 replied:
moved to after line 145. and dropped `context.n_ref--;` on line 144.

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,227 @@
+                r = metrics_call(p, event, varlinks[i], &context);
+                if (r < 0) {
+                        context.n_ref--;
+                        continue;
```
please do not silently ignore errors. systemd does a special patter when ignoring errors:

```
log_at_suitable_level(r, "Failed to ...., ignoring");
```

YapingLi04 replied:
added `log_error_errno()`

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,227 @@
+static int metrics_start_query(void) {
+        _cleanup_closedir_ DIR *d = NULL;
+        _cleanup_(sd_event_unrefp) sd_event *event = NULL;
+        int r, c;
```
I don't see `c` set anywhere

YapingLi04 replied:
`c` is being returned from `metrics_start_query()`

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,227 @@
+
+        r = sd_event_loop(event);
+        if (r < 0)
+                r = log_error_errno(r, "Failed to run event loop: %m");
```
use `RET_GATHER`

YapingLi04 replied:
@ikruglov did you mean to use `RET_GATHER` on line 146?

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,227 @@
+                        break;
+        }
+
+        r = sd_event_add_time_relative(event, NULL, CLOCK_MONOTONIC, 100, 0, timeout_callback, NULL);
```
/* foo= */ comments before NULL

ikruglov replied:
also, timeout should be per connection. Not a timer for entire event loop. Use `sd_varlink_set_relative_timeout()`

ikruglov commented on src/shared/metrics.c:
```diff
@@ -0,0 +1,235 @@
+        }
+
+        if (previous) {
+                r = metric_family_build_json_one(link, previous, /* more */ false);
```
/* more= */

ikruglov commented on src/shared/metrics.c:
```diff
@@ -0,0 +1,235 @@
+
+        r = sd_json_variant_new_unsigned(&v, value);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to allocate JSON string: %m");
```
copy-paste

YapingLi04 replied:
changed to `Failed to allocate JSON unsigned`

ikruglov commented on src/shared/metrics.c:
```diff
@@ -0,0 +1,235 @@
+        if (!FLAGS_SET(flags, SD_VARLINK_METHOD_MORE))
+                return sd_varlink_error(link, SD_VARLINK_ERROR_EXPECTED_MORE, NULL);
+
+        MetricFamilyContext ctx = {.link = link};
```
`.previous` still leaks

YapingLi04 replied:
this is fixed.

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,227 @@
+        if (r < 0)
+                return log_error_errno(r, "Failed to enable exit on SIGINT/SIGTERM: %m");
+
+        sd_varlink *varlinks[MAX_CONCURRENT_METRICS_SOCKETS] = {};
```
sd_varlink *varlinks[n_varlinks] = {};

YapingLi04 replied:
changed to dynamically allocated array.

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,227 @@
+        if (r < 0)
+                return log_debug_errno(r, "Failed to bind reply callback: %m");
+
+        r = sd_varlink_observe(vl, "io.systemd.Metrics.List", /* parameter */ NULL);
```
/* parameter= */

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,224 @@
+        fflush(stdout);
+
+        if (!FLAGS_SET(flags, SD_VARLINK_REPLY_CONTINUES) && (--context->n_ref <= 0))
+                (void) sd_event_exit(ASSERT_PTR(sd_varlink_get_event(link)), FLAGS_SET(flags, SD_VARLINK_REPLY_ERROR) ? EXIT_FAILURE: EXIT_SUCCESS);
```
@YapingLi04, check how `varlink_dispatch_reply()` is implemented.

There are several interesting things these.
1. `error_id` is only set together with `SD_VARLINK_REPLY_ERROR`. As result, the extra check for `SD_VARLINK_REPLY_ERROR`, given `if (error_id) {}`, is redundant.
2. There is also
```
        if (error && FLAGS_SET(flags, SD_VARLINK_REPLY_CONTINUES))
                goto invalid;
```

meaning that the callback will not get `SD_VARLINK_REPLY_ERROR` and `SD_VARLINK_REPLY_CONTINUES` at the same time.

To sum up. You should check for either `error_id` or `SD_VARLINK_REPLY_ERROR` for a error condition. You already do that. After that a simple check for `SD_VARLINK_REPLY_CONTINUES` is enough (see below)

```
if (!FLAGS_SET(flags, SD_VARLINK_REPLY_CONTINUES) && (--context->n_ref == 0))
     (void) sd_event_exit(ASSERT_PTR(sd_varlink_get_event(link)), EXIT_SUCCESS);
```

ikruglov replied:
Another thing. `n_ref` is unsigned. It cannot go below zero.

YapingLi04 replied:
Thanks for the reference! This is updated.

YapingLi04 reviewed (commented):
Addressed all comments and there should be no more memory leaks.

I have one question on the comment of use`RET_GATHER` on report.c

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,231 @@
+
+        if (error_id) {
+                bool disconnect = streq(error_id, SD_VARLINK_ERROR_DISCONNECTED);
+                bool timout = streq(error_id, SD_VARLINK_ERROR_TIMEOUT);
```
timeout

YapingLi04 replied:
corrected spelling

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,231 @@
+                else
+                        log_error("Varlink error: %s", error_id);
+
+                goto finish;
```
good catch

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,231 @@
+                bool disconnect = streq(error_id, SD_VARLINK_ERROR_DISCONNECTED);
+                bool timout = streq(error_id, SD_VARLINK_ERROR_TIMEOUT);
+                if (disconnect)
+                        log_info("Varlink disconnected.");
```
nit: ideally, you should log which exact connection disconnected/timeout/errors.

YapingLi04 replied:
I don't see a straightforward way to get the socket address from a given varlink.

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,231 @@
+                if (errno == ENOENT)
+                        return log_error_errno(-ESRCH, "No metrics sockets found: %m");
+
+                return log_error_errno(-errno, "Failed to open directory: %m");
```
log_error_errno("Failed to open directory %s: %m", RUN_SYSTEMD_METRICS_PATH)

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,231 @@
+        d = opendir(RUN_SYSTEMD_METRICS_PATH);
+        if (!d) {
+                if (errno == ENOENT)
+                        return log_error_errno(-ESRCH, "No metrics sockets found: %m");
```
multiple of things:
1. no `-ESRCH`, just `ESRCH` if you want ESRCH
2. if you want to use ESRCH, it should be `SYNTHETIC_ERRNO(ESRCH)`. Since you come up with `ESRCH` out of blue
3. I'd just use `ENOENT`

YapingLi04 replied:
Updated to use `ENOENT`

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,231 @@
+                context.n_ref++;
+
+                if (++i >= MAX_CONCURRENT_METRICS_SOCKETS) {
+                        log_warning("Too many concurrent metrics sockets, skipping %s", p);
```
log_warning("Too many concurrent metrics sockets, skipping %s", p);

a bit misleading because you break

log_warning("Too many concurrent metrics sockets, stop iterating");

seems a bit better. But this is just a taste question

YapingLi04 replied:
Updated to log_warning("Too many concurrent metrics sockets, stop iterating");

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,231 @@
+
+        r = sd_event_get_exit_code(event, &c);
+        if (r < 0)
+                r = log_error_errno(r, "Failed to get exit code: %m");
```
error handling doesn't make sense to me here. I guess you can just use:
```
         r = sd_event_loop(event);
        if (r < 0)
                return log_error_errno(r, "Failed to run event loop: %m");
        return 0;
```

YapingLi04 replied:
updated the code and dropped `c`.

ikruglov reviewed (changes requested):
@YapingLi04, the PR is taking shape! Some minor comments.
Let's ask @DaanDeMeyer to give it another round.

ikruglov commented on src/report/report.c:
```diff
@@ -0,0 +1,227 @@
+        return 0;
+}
+
+static int metrics_call(const char *path, sd_event *event, sd_varlink **vl, Context *context) {
```
forgot to mention, you still need to change `vl`->`ret` (per coding style) and `link`->`vl` for convinience.

ikruglov commented:
@YapingLi04, I believe your most recent change introduced a memleak.

ikruglov commented:
@YapingLi04, you latest push with a rebase made this PR unreviewable. You need to update `main` with the same rebase to be able to see only your changes here.

YapingLi04 commented:
I did push main with the same rebase.

ikruglov commented on src/shared/metrics.c:
```diff
@@ -0,0 +1,290 @@
+
+        STRV_FOREACH_PAIR(fk, fv, field_pairs) {
+                n_pairs++;
+        }
```
use `strv_length()`

ikruglov replied:
also, you need a short cut when empty or null field_pairs

ikruglov replied:
also assert that field_pairs have even amount of items

YapingLi04 replied:
@ikruglov how do you short circuit when using `strv_length()`?

YapingLi04 replied:
Ah, I see what you mean. Updated to return early when field_pairs is null.

ikruglov commented on src/shared/metrics.c:
```diff
@@ -0,0 +1,290 @@
+        STRV_FOREACH_PAIR(key, value, field_pairs) {
+                r = sd_json_variant_new_string(&array[n++], *key);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to create key variant: %m");
```
`log_debug_errno()`, it's a helper function

ikruglov commented on src/shared/metrics.c:
```diff
@@ -0,0 +1,290 @@
+
+                r = sd_json_variant_new_string(&array[n++], *value);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to create value variant: %m");
```
ditto

ikruglov commented on src/shared/metrics.c:
```diff
@@ -0,0 +1,290 @@
+
+        r = sd_json_variant_new_object(&w, array, n);
+        if (r < 0)
+                return log_error_errno(r, "Failed to allocate JSON object: %m");
```
ditto

ikruglov commented on src/shared/metrics.c:
```diff
@@ -0,0 +1,290 @@
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *w = NULL;
+        size_t n_pairs = 0;
+        int r;
+
```
assert(v) earlier

ikruglov commented on src/shared/varlink-io.systemd.Metrics.c:
```diff
@@ -0,0 +1,68 @@
+                SD_VARLINK_FIELD_COMMENT("A int value type"),
+                SD_VARLINK_DEFINE_FIELD(intVal, SD_VARLINK_INT, SD_VARLINK_NULLABLE),
+                SD_VARLINK_FIELD_COMMENT("A boolean value type"),
+                SD_VARLINK_DEFINE_FIELD(boolVal, SD_VARLINK_BOOL, SD_VARLINK_NULLABLE));
```
str, int, bool I'd say. Make it shorter
